physics/vector: Improved speed of Point/Frame partial_velocity methods

ricdigi opened the pull request:
#### References to other Issues or PRs
- Issue #25070: 
The issue highlights inefficiencies in the mechanics module, particularly where diff() and jacobian() are used to derive linear coefficients in expressions assumed to be linear.
- PR #26367: 
This PR introduced a speed improvement in the partial_velocity function in physics.vector


#### Brief description of what is fixed or changed

The optimised method to compute partial_velocity which leverages the linearity of velocities in the generalised coordinates, already discussed in #26367, has been now implemented in the partial_velocity method of Point and Frame classes.

#### Release Notes


<!-- BEGIN RELEASE NOTES -->
* physics.vector
  * Improved speed of Point/Frame partial_velocity methods

<!-- END RELEASE NOTES -->

moorepants commented:
Do we have duplicate code in 3 places now with this PR?

ricdigi commented:
> Do we have duplicate code in 3 places now with this PR?

I wanted to use the same function, but I noticed that the inside the two partial_velocity methods the partial_velocity() function present in physics/vector/functions.py was not used, instead there was another implementation. Thus, I duplicated the code. 

However I can modify it again to use the partial_velocity() function present in physics/vector/functions.py

moorepants commented:
If it is possible to have one location for the code and then reuse it in other places, that would be best. If you have time and the inclination to make it like that, please do so. Otherwise we can merge this.

ricdigi committed:
physics/vector: Improved speed of Point/Frame partial_velocity methods
This commit optimizes the partial_velocity method of Point and Frame classes,
leveraging the linearity relationship between velocities and generalized speeds.
It removes duplicate code by utilizing the partial_velocity function in
physics.vector.functions updated with PR:#26367

Related issue: #25070
The issue is about the potential speed gains in the mechanics module
obtained by replacing more generic, computation-intensive operations like
jacobian() with other methods that more efficiently extract linear coefficients.

ricdigi commented:
I have recently revised the methods to utilize the updated partial_velocity function from #26367. Additionally, I completely removed the previous commit by performing a force push for a cleaner look. Is this approach acceptable, or should I have handled it differently?

moorepants commented:
It's fine, but I don't recommend force pushing. When a review occurs in a PR, if you delete old commits you can't follow the review comments anymore because the code is gone. Once you push git commits to the public you shouldn't change them, except for some special cases.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -1558,8 +1558,10 @@ def partial_velocity(self, frame, *gen_speeds):
 
-        partials = [self.ang_vel_in(frame).diff(speed, frame, var_in_dcm=False)
-                    for speed in gen_speeds]
+        from sympy.physics.vector.functions import partial_velocity
```
Are these imports here because of some circular dependency issue? If not, then they should be at the top of the file with the other imports. If it is a circular dependency then we need to rethink where the primary function should be stored.

ricdigi replied:
> Are these imports here because of some circular dependency issue? If not, then they should be at the top of the file with the other imports. If it is a circular dependency then we need to rethink where the primary function should be stored.

Yes, I put them inside because I was getting a circular dependency issue, but only for the frame class. However for consistency I put the imports inside the methods definition for both classes. What's the procedure in these cases?

moorepants replied:
We can keep the imports like this if needed, but identifying why it occurs and fixing it can be better. I guess the partial_velocity function depends on things in frame.

moorepants replied:
Procedure is to only put imports within functions/methods if you have to because you can't fix the circular dependency or maybe if you are trying to somehow lazy load a module for specific import performance reasons.

moorepants replied:
I guess this is because functions.py imports things from frame and not point.

ricdigi replied:
Maybe the circular dependency issue is the reason why the code was "duplicated" in both classes and partial_velocity() from functions.py was not used

moorepants replied:
Possibly!

moorepants replied:
You can also go back to that. We could just open an issue about the circular dependencies.

tjstienstra replied:
> You can also go back to that. We could just open an issue about the circular dependencies.

Opened #26411

A possible fix for these circular imports would be to create a `Vector.partial_diff(self, vars, linear=False)` method.
